java-checks/src/main/java/org/sonar/java/checks/spring/SpelExpressionCheck.java

```diff
@@ -0,0 +1,246 @@
+      tokenStart.line(),
+      tokenStart.columnOffset() + error.rangeEnd
+    );
+
+    var analyzerMessage = new AnalyzerMessage(this, context.getInputFile(), textSpan, error.getMessage(), 0);
+    ((DefaultJavaFileScannerContext) context).reportIssue(analyzerMessage);
+  }
+
+  private static void parseStringContents(String content) {
+    var i = 0;
+    while (i < content.length()) {
+      var c = content.charAt(i);
+      switch (c) {
+        case '$':
+          i = parseDelimitersAndContents(content, i + 1, i + 1, SpelExpressionCheck::checkValidPropertyPlaceholder);
+          break;
+        case '#':
+          i = parseDelimitersAndContents(content, i + 1, i + 1, SpelExpressionCheck::checkValidSpelExpression);
+          break;
+        default:
+          i++;
+          break;
+      }
+    }
+  }
+
+  private static int parseDelimitersAndContents(
+    String value,
+    int startIndex,
+    int rangeStart,
+    ObjIntConsumer<String> parseContents) {
+    if (startIndex == value.length()) {
+      return startIndex;
+    }
+    var endIndex = parseDelimiterBraces(value, startIndex, rangeStart);
+    if (endIndex == startIndex) {
+      return endIndex;
+    }
+    var contents = value.substring(startIndex + 1, endIndex - 1);
+    parseContents.accept(contents, rangeStart);
+    return endIndex;
+  }
+
+  private static int parseDelimiterBraces(String value, int startIndex, int rangeStart) throws SyntaxError {
+    if (value.charAt(startIndex) != '{') {
+      return startIndex;
+    }
+    var i = startIndex + 1;
+    var openCount = 1;
+
+    while (i < value.length()) {
+      var c = value.charAt(i);
+      i++;
+      switch (c) {
+        case '{':
+          openCount++;
+          break;
+        case '}':
+          openCount--;
+          if (openCount == 0) {
+            return i;
+          }
+          break;
+        default:
+          break;
+      }
+    }
+
+    // +1 because of prefix `$` or `#`
+    var rangeEnd = rangeStart + i - startIndex + 1;
+    throw new SyntaxError("Add missing '}' for this property placeholder or SpEL expression.", rangeStart, rangeEnd);
+  }
+
+  private static void checkValidPropertyPlaceholder(String placeholder, int rangeStart) {
+    if (!isValidPropertyPlaceholder(placeholder, rangeStart)) {
+      // +3 because of delimiter `#{` and `}`
+      var rangeEnd = rangeStart + placeholder.length() + 3;
+      throw new SyntaxError("Correct this malformed property placeholder.", rangeStart, rangeEnd);
+    }
+  }
+
+  private static boolean isValidPropertyPlaceholder(String placeholder, int rangeStart) {
+    var startIndex = 0;
+    var endIndex = placeholder.indexOf(':');
+
+    while (endIndex != -1) {
+      var segment = placeholder.substring(startIndex, endIndex);
+      if (!isValidPropertyPlaceholderSegment(segment, rangeStart + startIndex)) {
+        return false;
+      }
+      startIndex = endIndex + 1;
+      endIndex = placeholder.indexOf(':', startIndex);
```

Comment:
My understanding is that the column token marks an alternative in SpEL syntax. In practice, is there a limit to the number of alternatives we can encounter?

Reply:
Probably not: I have not found a single example that has more than one `:` in a property placeholder. However, I have also not found an explicit specification of what is allowed and what not, or even a grammar for property placeholders. Best doc I have found so far is the API spec of the Spring PropertyPlaceholderConfigurer.

Reply:
Indeed, from this [outdated grammar](https://github.com/spring-projects/spring-framework/blob/2a268f2250a1016c0c1444a9a4a467c8f2b71d32/spring-expression/src/main/java/org/springframework/expression/spel/generated/SpringExpressions.g#L47C1-L47C1), I am not sure that there is a limit on the number of alternatives